Replace set_last_error with set_last_error_and_return

noahmbright opened the pull request:
Took care of the simple patterns. Other patterns involved setting an error and then using `write_int` or setting metadata and returning -1. Unsure if those are in the scope of this change

Looks like this has conflicts with #3779, so I can update when how to handle that is decided. 

Part of https://github.com/rust-lang/miri/issues/3930.

oli-obk commented on src/shims/unix/fs.rs:
Where did the stray formatting changes come from?

oli-obk commented on src/shims/unix/fd.rs:
```diff
@@ -554,8 +554,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
         let this = self.eval_context_mut();
         let ebadf = this.eval_libc("EBADF");
-        this.set_last_error(ebadf)?;
-        Ok((-1).into())
+        return this.set_last_error_and_return_i32(ebadf);
```
Use `LibcError("EBADF")`

oli-obk replied:
Also in other places

oli-obk reviewed (changes requested):
Please revert the formatting changes unless they are from `./miri fmt` (you may need to configure your vscode correctly with the settings from `CONTRIBUTING.md`)

Also use `LibcError("foobar")` everywhere while you're doing this change.

oli-obk commented:
Ignore my PR, i'll rebase at some point

RalfJung commented on src/shims/unix/fs.rs:
```diff
@@ -600,8 +594,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             SeekFrom::End(i64::try_from(offset).unwrap())
         } else {
-            this.set_last_error(LibcError("EINVAL"))?;
-            return Ok(Scalar::from_i64(-1));
```
This returns an `i64`, not an `i32`, so your patch introduces an ICE here! So please add a new `set_last_error_and_return_i64` instead and use that here, or leave this to a future PR.

Please make sure to always maintain the original type, it's quite important to not screw this up and because this is "dynamic" (it's types of the interpreted program, not of the interpreter), even `./miri check` will not catch this.

RalfJung commented on src/shims/unix/fs.rs:
```diff
@@ -590,8 +585,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             if offset < 0 {
                 // Negative offsets return `EINVAL`.
-                this.set_last_error(LibcError("EINVAL"))?;
-                return Ok(Scalar::from_i64(-1));
```
More incorrect `i64` -> `i32` changes

RalfJung commented on src/shims/unix/fs.rs:
```diff
@@ -1026,8 +1013,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             this.reject_in_isolation("`opendir`", reject_with)?;
             let eacc = this.eval_libc("EACCES");
-            this.set_last_error(eacc)?;
-            return Ok(Scalar::null_ptr(this));
```
This returned a null pointer before and now you made it return -1. That's definitely wrong, please keep the old code here.

tiif commented on src/shims/unix/fs.rs:
```diff
@@ -1441,8 +1419,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             let eacc = this.eval_libc("EACCES");
-            this.set_last_error(eacc)?;
-            return Ok(-1);
+            return Ok(this.set_last_error_and_return_i32(eacc));
```
It seems like previously it returns ``Ok(-1)`` but now it returns ``Ok(Ok(-1))``?

RalfJung commented:
Not sure why the comments above have all been marked as resolved -- in the PR they are definitely still unresolved.

@rustbot author

noahmbright commented:
> Not sure why the comments above have all been marked as resolved -- in the PR they are definitely still unresolved.

Yes, sorry for the wait. How do you want to handle the `InterpResult<'tcx, i64>` vs . `InterpResult<'tcx, Scalar>` issue in my comment above? Once I fix that I'll push the update.

RalfJung commented:
Which comment?

RalfJung commented on src/shims/unix/fs.rs:
```diff
@@ -1441,8 +1419,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             let eacc = this.eval_libc("EACCES");
-            this.set_last_error(eacc)?;
-            return Ok(-1);
+            return Ok(this.set_last_error_and_return_i32(eacc));
```
For now just change this to `this.set_last_error(LibcError("EACCES"))?;` but leave the `return` alone.

noahmbright commented:
Again sorry about the wait and the careless mistakes, but I added two commits, one a small formatting fix, and in the other I added `set_last_error_and_return_i64` and fixed the type issues I introduced. Diff is smaller now too so hopefully this will be easier to re-review.

RalfJung commented:
There are unfortunately conflicts with the master branch, could you take care of them?
Then write @`rustbot ready`  to initiate the next round of review.

noahmbright commented:
> There are unfortunately conflicts with the master branch, could you take care of them with a rebase? Then write @`rustbot ready` to initiate the next round of review.

Of course. All the conflicts are something like this:
```
<<<<<<< unix_shims
            return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);
=======
            this.set_last_error(ErrorKind::PermissionDenied)?;
            return interp_ok(Scalar::from_i32(-1));
>>>>>>> master
```

I can wrap `interp_ok(this.set_last_error_and_return_*(e))` where these all occur, and we'll have an `interp_ok(InterpResult<'tcx, Scalar>)`. Does that sound alright for the resolution?

RalfJung commented:
`return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);` looks like the right resolution for that example.

noahmbright commented:
> `return this.set_last_error_and_return_i32(ErrorKind::PermissionDenied);` looks like the right resolution for that example.

I agree, after seeing the new changes while rebasing. @rustbot ready

RalfJung commented on src/shims/unix/linux/epoll.rs:
```diff
@@ -375,8 +370,7 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
             // Remove epoll_event_interest from interest_list.
             let Some(epoll_interest) = interest_list.remove(&epoll_key) else {
                 let enoent = this.eval_libc("ENOENT");
```
This could use `LibcError`

RalfJung reviewed (commented):
Thanks, this all looks correct but there is one spot where we can make it a bit more compact. :)

Please also squash the commits (with `git rebase --keep-base`).

RalfJung commented:
@rustbot author

noahmbright commented:
> Thanks, this all looks correct but there is one spot where we can make it a bit more compact. :)

Good catch. Thank you for the close reviews, this should have been much simpler. If I'm still welcome, I'll make sure my next PR goes more smoothly

RalfJung commented:
Looks good, thanks for the PR!
@bors r+

> Thank you for the close reviews, this should have been much simpler. If I'm still welcome, I'll make sure my next PR goes more smoothly

Sure thing, hickups happen. :) Just be careful and ask us rather than making guesses if you're not sure about something. :D When submitting a PR, it is a very good idea to list which things in the PR you are least sure about -- that greatly helps focus the review effort.

bors commented:
:pushpin: Commit ca5c26df2960fa993898b8fc0527ae88b989648f has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/miri) for this repository.

<!-- @bors r=RalfJung ca5c26df2960fa993898b8fc0527ae88b989648f -->
<!-- homu: {"type":"Approved","sha":"ca5c26df2960fa993898b8fc0527ae88b989648f","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/miri"} -->

bors commented:
:hourglass: Testing commit ca5c26df2960fa993898b8fc0527ae88b989648f with merge 0a916c547074d03804245d1d89d2d84417d1e577...
<!-- homu: {"type":"BuildStarted","head_sha":"ca5c26df2960fa993898b8fc0527ae88b989648f","merge_sha":"0a916c547074d03804245d1d89d2d84417d1e577"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/actions/runs/11437977563/job/31819818526)
Approved by: RalfJung
Pushing 0a916c547074d03804245d1d89d2d84417d1e577 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/actions/runs/11437977563/job/31819818526"},"merge_sha":"0a916c547074d03804245d1d89d2d84417d1e577"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 1 review requesting changes by reviewers with write access.

RalfJung commented:
That's odd... @rust-lang/infra bors has a problem if we leave a "changes required" review, and then the changes are done? CI is green but the PR has not been pushed to master.

RalfJung reviewed (approved):
Let's see if this helps github.

RalfJung commented:
@bors retry

Kobzol commented:
Wtf, I never saw this error before.. Anyway, what do you think about switching miri from bors to merge queues? xD

RalfJung commented:
@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors commented:
:pushpin: Commit ca5c26df2960fa993898b8fc0527ae88b989648f has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/miri) for this repository.

<!-- @bors r=RalfJung ca5c26df2960fa993898b8fc0527ae88b989648f -->
<!-- homu: {"type":"Approved","sha":"ca5c26df2960fa993898b8fc0527ae88b989648f","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/miri"} -->

RalfJung commented:
> Wtf, I never saw this error before.. Anyway, what do you think about switching miri from bors to merge queues? xD

Yeah someone just needs to put in the work for that and I don't currently have the time.^^

bors commented:
:hourglass: Testing commit ca5c26df2960fa993898b8fc0527ae88b989648f with merge 652e860cd660fa3b1d1d2bbb5a59a6b6760ab6af...
<!-- homu: {"type":"BuildStarted","head_sha":"ca5c26df2960fa993898b8fc0527ae88b989648f","merge_sha":"652e860cd660fa3b1d1d2bbb5a59a6b6760ab6af"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/actions/runs/11439043433/job/31823066346)
Approved by: RalfJung
Pushing 652e860cd660fa3b1d1d2bbb5a59a6b6760ab6af to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/actions/runs/11439043433/job/31823066346"},"merge_sha":"652e860cd660fa3b1d1d2bbb5a59a6b6760ab6af"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 1 review requesting changes and 1 approving review by reviewers with write access.

RalfJung commented:
@oli-obk I think we need to you to send an "approving" review or so, otherwise this PR can't land...

Is that really how github reviews work? Nobody else can approve a PR after a team member left a "changes required" review? That's a complete disaster.

oli-obk commented:
you should be able to dismiss my review in the github ui (i approved it now, but for the future)

@bors r+

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors commented:
:pushpin: Commit ca5c26df2960fa993898b8fc0527ae88b989648f has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/miri) for this repository.

<!-- @bors r=oli-obk ca5c26df2960fa993898b8fc0527ae88b989648f -->
<!-- homu: {"type":"Approved","sha":"ca5c26df2960fa993898b8fc0527ae88b989648f","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/miri"} -->

RalfJung commented:
Where is the "dismiss review" button?

RalfJung commented:
Ah I think I found it. That's quite well-hidden...

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Changes must be made through a pull request.
